File: Tutorials/Codes/Modern_Control_Course/Chapter27/Lesson4/Chapter27_Lesson4.java
/*
Chapter27_Lesson4.java
State-space disturbance rejection using integral action.

Compile and run:
    javac Chapter27_Lesson4.java
    java Chapter27_Lesson4
*/

import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Arrays;

public class Chapter27_Lesson4 {
    static double[][] multiply(double[][] A, double[][] B) {
        int m = A.length;
        int n = B[0].length;
        int p = B.length;
        double[][] C = new double[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < p; k++) {
                    C[i][j] += A[i][k] * B[k][j];
                }
            }
        }
        return C;
    }

    static double[][] add(double[][] A, double[][] B) {
        double[][] C = new double[A.length][A[0].length];
        for (int i = 0; i < A.length; i++) {
            for (int j = 0; j < A[0].length; j++) {
                C[i][j] = A[i][j] + B[i][j];
            }
        }
        return C;
    }

    static double[][] scale(double[][] A, double alpha) {
        double[][] C = new double[A.length][A[0].length];
        for (int i = 0; i < A.length; i++) {
            for (int j = 0; j < A[0].length; j++) {
                C[i][j] = alpha * A[i][j];
            }
        }
        return C;
    }

    static double[][] identity(int n) {
        double[][] I = new double[n][n];
        for (int i = 0; i < n; i++) I[i][i] = 1.0;
        return I;
    }

    static double[][] inverse(double[][] A) {
        int n = A.length;
        double[][] aug = new double[n][2 * n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) aug[i][j] = A[i][j];
            aug[i][n + i] = 1.0;
        }

        for (int col = 0; col < n; col++) {
            int pivot = col;
            for (int row = col + 1; row < n; row++) {
                if (Math.abs(aug[row][col]) > Math.abs(aug[pivot][col])) {
                    pivot = row;
                }
            }
            double[] temp = aug[col];
            aug[col] = aug[pivot];
            aug[pivot] = temp;

            double div = aug[col][col];
            if (Math.abs(div) < 1e-12) throw new RuntimeException("Singular matrix");
            for (int j = 0; j < 2 * n; j++) aug[col][j] /= div;

            for (int row = 0; row < n; row++) {
                if (row == col) continue;
                double factor = aug[row][col];
                for (int j = 0; j < 2 * n; j++) {
                    aug[row][j] -= factor * aug[col][j];
                }
            }
        }

        double[][] inv = new double[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) inv[i][j] = aug[i][n + j];
        }
        return inv;
    }

    static double[][] column(double a, double b, double c) {
        return new double[][]{{a}, {b}, {c}};
    }

    static double[][] controllability(double[][] A, double[][] B) {
        double[][] AB = multiply(A, B);
        double[][] A2B = multiply(A, AB);
        return new double[][]{
                {B[0][0], AB[0][0], A2B[0][0]},
                {B[1][0], AB[1][0], A2B[1][0]},
                {B[2][0], AB[2][0], A2B[2][0]}
        };
    }

    static double[][] desiredPolynomialMatrix(double[][] A) {
        // Desired poles: -2, -2.5, -3
        // p(s) = s^3 + 7.5 s^2 + 18.5 s + 15
        double[][] A2 = multiply(A, A);
        double[][] A3 = multiply(A2, A);
        return add(add(add(A3, scale(A2, 7.5)), scale(A, 18.5)), scale(identity(3), 15.0));
    }

    static double[] ackermannGain(double[][] A, double[][] B) {
        double[][] ctrb = controllability(A, B);
        double[][] phiA = desiredPolynomialMatrix(A);
        double[][] last = new double[][]{{0.0, 0.0, 1.0}};
        double[][] K = multiply(multiply(last, inverse(ctrb)), phiA);
        return K[0];
    }

    static double[] dynamics(double[] xa, double[] K, double d0) {
        double x1 = xa[0];
        double x2 = xa[1];
        double xi = xa[2];
        double u = -(K[0] * x1 + K[1] * x2 + K[2] * xi);
        return new double[]{
                x2,
                -2.0 * x1 - 0.6 * x2 + u + d0,
                x1
        };
    }

    static double[] addScaled(double[] x, double[] dx, double h) {
        return new double[]{x[0] + h * dx[0], x[1] + h * dx[1], x[2] + h * dx[2]};
    }

    public static void main(String[] args) throws IOException {
        double[][] Aaug = new double[][]{
                {0.0, 1.0, 0.0},
                {-2.0, -0.6, 0.0},
                {1.0, 0.0, 0.0}
        };
        double[][] Baug = column(0.0, 1.0, 0.0);
        double[] K = ackermannGain(Aaug, Baug);
        System.out.println("Kaug = " + Arrays.toString(K));

        double d0 = 0.5;
        double h = 0.001;
        double tf = 10.0;
        double[] xa = new double[]{0.4, 0.0, 0.0};

        try (PrintWriter out = new PrintWriter(new FileWriter("Chapter27_Lesson4_java_response.csv"))) {
            out.println("t,x1,x2,xi,u,y");
            int steps = (int) (tf / h);
            for (int k = 0; k <= steps; k++) {
                double t = k * h;
                double u = -(K[0] * xa[0] + K[1] * xa[1] + K[2] * xa[2]);
                out.printf("%.10f,%.10f,%.10f,%.10f,%.10f,%.10f%n", t, xa[0], xa[1], xa[2], u, xa[0]);

                double[] k1 = dynamics(xa, K, d0);
                double[] k2 = dynamics(addScaled(xa, k1, 0.5 * h), K, d0);
                double[] k3 = dynamics(addScaled(xa, k2, 0.5 * h), K, d0);
                double[] k4 = dynamics(addScaled(xa, k3, h), K, d0);
                for (int i = 0; i < 3; i++) {
                    xa[i] += (h / 6.0) * (k1[i] + 2.0 * k2[i] + 2.0 * k3[i] + k4[i]);
                }
            }
        }

        System.out.println("Final y = " + xa[0]);
        System.out.println("CSV written to Chapter27_Lesson4_java_response.csv");
    }
}

File: Tutorials/Codes/Modern_Control_Course/Chapter27/Lesson4/Chapter27_Lesson4.cpp
/*
Chapter27_Lesson4.cpp
State-space disturbance rejection using integral action.

Dependency:
    Eigen 3
Compile example:
    g++ -std=c++17 Chapter27_Lesson4.cpp -I /path/to/eigen -O2 -o Chapter27_Lesson4
*/

#include <Eigen/Dense>
#include <fstream>
#include <iomanip>
#include <iostream>

using Matrix3 = Eigen::Matrix3d;
using Vector3 = Eigen::Vector3d;
using Row3 = Eigen::RowVector3d;

Matrix3 desiredPolynomialMatrix(const Matrix3& Aaug) {
    // Desired poles: -2, -2.5, -3
    // p(s) = s^3 + 7.5 s^2 + 18.5 s + 15
    return Aaug * Aaug * Aaug + 7.5 * Aaug * Aaug + 18.5 * Aaug + 15.0 * Matrix3::Identity();
}

Row3 ackermannGain(const Matrix3& Aaug, const Vector3& Baug) {
    Matrix3 ctrb;
    ctrb.col(0) = Baug;
    ctrb.col(1) = Aaug * Baug;
    ctrb.col(2) = Aaug * Aaug * Baug;

    Row3 last;
    last << 0.0, 0.0, 1.0;

    Matrix3 phiA = desiredPolynomialMatrix(Aaug);
    return last * ctrb.inverse() * phiA;
}

Vector3 dynamics(const Vector3& xa, const Row3& Kaug, double d0) {
    // Plant: x1_dot = x2, x2_dot = -2 x1 -0.6 x2 + u + d0, xi_dot = x1
    double x1 = xa(0);
    double x2 = xa(1);
    double xi = xa(2);
    double u = -(Kaug(0) * x1 + Kaug(1) * x2 + Kaug(2) * xi);

    Vector3 dx;
    dx << x2,
          -2.0 * x1 - 0.6 * x2 + u + d0,
          x1;
    return dx;
}

int main() {
    Matrix3 Aaug;
    Aaug << 0.0, 1.0, 0.0,
           -2.0, -0.6, 0.0,
            1.0, 0.0, 0.0;

    Vector3 Baug;
    Baug << 0.0, 1.0, 0.0;

    Row3 Kaug = ackermannGain(Aaug, Baug);
    std::cout << "Kaug = " << Kaug << std::endl;

    const double d0 = 0.5;
    const double h = 0.001;
    const double tf = 10.0;
    Vector3 xa;
    xa << 0.4, 0.0, 0.0;

    std::ofstream file("Chapter27_Lesson4_cpp_response.csv");
    file << "t,x1,x2,xi,u,y\n";

    for (int k = 0; k <= static_cast<int>(tf / h); ++k) {
        double t = k * h;
        double u = -(Kaug(0) * xa(0) + Kaug(1) * xa(1) + Kaug(2) * xa(2));
        double y = xa(0);
        file << std::setprecision(10) << t << "," << xa(0) << "," << xa(1) << ","
             << xa(2) << "," << u << "," << y << "\n";

        Vector3 k1 = dynamics(xa, Kaug, d0);
        Vector3 k2 = dynamics(xa + 0.5 * h * k1, Kaug, d0);
        Vector3 k3 = dynamics(xa + 0.5 * h * k2, Kaug, d0);
        Vector3 k4 = dynamics(xa + h * k3, Kaug, d0);
        xa += (h / 6.0) * (k1 + 2.0 * k2 + 2.0 * k3 + k4);
    }

    std::cout << "Final y = " << xa(0) << std::endl;
    std::cout << "CSV written to Chapter27_Lesson4_cpp_response.csv" << std::endl;
    return 0;
}
